common/headtracker/head_tracker.go

```diff
@@ -335,12 +341,58 @@ func (ht *headTracker[HTH, S, ID, BLOCK_HASH]) backfillLoop() {
 	}
 }
 
+// LatestAndFinalizedBlock - returns latest and latest finalized blocks.
+// NOTE: Returns latest finalized block as is, ignoring the FinalityTagBypass feature flag.
+// TODO: BCI-3321 use cached values instead of making RPC requests
+func (ht *headTracker[HTH, S, ID, BLOCK_HASH]) LatestAndFinalizedBlock(ctx context.Context) (latest, finalized HTH, err error) {
+	latest, err = ht.client.HeadByNumber(ctx, nil)
+	if err != nil {
+		err = fmt.Errorf("failed to get latest block: %w", err)
+		return
+	}
+
+	if !latest.IsValid() {
+		err = fmt.Errorf("expected latest block to be valid")
+		return
+	}
+
+	finalized, err = ht.calculateLatestFinalized(ctx, latest, false)
+	if err != nil {
+		err = fmt.Errorf("failed to calculate latest finalized block: %w", err)
+		return
+	}
+	if !finalized.IsValid() {
+		err = fmt.Errorf("expected finalized block to be valid")
+		return
+	}
+
+	return
+}
+
+func (ht *headTracker[HTH, S, ID, BLOCK_HASH]) getHeadAtHeight(ctx context.Context, chainHeadHash BLOCK_HASH, blockHeight int64) (HTH, error) {
+	chainHead := ht.headSaver.Chain(chainHeadHash)
+	if chainHead.IsValid() {
+		// check if provided chain contains a block of specified height
+		headAtHeight, err := chainHead.HeadAtHeight(blockHeight)
+		if err == nil {
+			// we are forced to reload the block due to type mismatched caused by generics
+			hthAtHeight := ht.headSaver.Chain(headAtHeight.BlockHash())
+			// ensure that the block was not removed from the chain by another goroutine
+			if hthAtHeight.IsValid() {
+				return hthAtHeight, nil
+			}
+		}
+	}
+
+	return ht.client.HeadByNumber(ctx, big.NewInt(blockHeight))
+}
+
 // calculateLatestFinalized - returns latest finalized block. It's expected that currentHeadNumber - is the head of
 // canonical chain. There is no guaranties that returned block belongs to the canonical chain. Additional verification
 // must be performed before usage.
-func (ht *headTracker[HTH, S, ID, BLOCK_HASH]) calculateLatestFinalized(ctx context.Context, currentHead HTH) (latestFinalized HTH, err error) {
-	if ht.config.FinalityTagEnabled() && !ht.htConfig.FinalityTagBypass() {
-		latestFinalized, err = ht.client.LatestFinalizedBlock(ctx)
+func (ht *headTracker[HTH, S, ID, BLOCK_HASH]) calculateLatestFinalized(ctx context.Context, currentHead HTH, finalityTagBypass bool) (HTH, error) {
+	if ht.config.FinalityTagEnabled() && !finalityTagBypass {
```

Comment:
`calculateLatestFinalized` is a method of headTracker and it already has access to its configs. Instead of passing it through the function params we should access it through headTracker's configs.
```suggestion
	if ht.config.FinalityTagEnabled() && !ht.htConfig.FinalityTagBypass() {
```

Reply:
We need to use variable instead of config option to pass `false` when calling `calculateLatestFinalized` from `LatestAndFinalizedBlock`. 
`LatestAndFinalizedBlock` is used by LogPoller and we need to ensure that it's not affected by the `FinalityTagBypass = true`.